app/models/status_reaction.rb

```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+
+# == Schema Information
+#
+# Table name: status_reactions
+#
+#  id              :bigint(8)        not null, primary key
+#  account_id      :bigint(8)        not null
+#  status_id       :bigint(8)        not null
+#  name            :string           default(""), not null
+#  custom_emoji_id :bigint(8)
+#  created_at      :datetime         not null
+#  updated_at      :datetime         not null
+#
+class StatusReaction < ApplicationRecord
+  belongs_to :account
+  belongs_to :status, inverse_of: :status_reactions
+  belongs_to :custom_emoji, optional: true
+
+  has_one :notification, as: :activity, dependent: :destroy
+
+  validates :name, presence: true
+  validates_with StatusReactionValidator
+
+  before_validation :set_custom_emoji
+
+  private
+
+  def set_custom_emoji
+    self.custom_emoji = CustomEmoji.find_by(shortcode: name, domain: account.domain) if name.blank?
+  end
```

Comment:
This doesn't make any sense to me.
I see that this comes from announcement reactions, but there it checks for `if name.present?` instead.
Right now it just tries to find a custom emoji with no shortcode, when there is no name given.
Though I still have no idea what this is for.

Reply:
```suggestion
  def set_custom_emoji
    self.custom_emoji = CustomEmoji.find_by(disabled: false, shortcode: name, domain: account.domain) if name.present?
  end
```

Oh I see what this is for.
This is supposed to prevent reacting with a disabled emoji and also sets custom_emoji when a custom emoji shortcode is given, but no custom_emoji object.

Reply:
Hmmm... Once this change is applied, you can no longer react with foreign custom emoji when such a reaction already existed on the status.

Reply:
Ah, yeah, because it looks for an emoji with the reactor's domain :/
It's something I unfortunately can't test for, but is the name passed as `shortcode@domain.tld` to this? In that case, it also should split between name and domain

Reply:
True. In fact, I've just observed incorrect behaviour happening due to account.domain being used...

Reply:
It seems your original idea with this change was correct. We just need a different value for domain.

Reply:
```suggestion
  def set_custom_emoji
    domain = custom_emoji.domain if custom_emoji.present?
    self.custom_emoji = CustomEmoji.find_by(disabled: false, shortcode: name, domain: domain) if name.present?
  end
```

This seems to fix not being able to react with remote custom emojis.
Though for some reason, you're unable to unreact with them, but I doubt it's this change, as it's not called for that.
I'm investigating.

Reply:
Ah, not being able to unreact was fixed by 0f96d0fbb7b9c76329d2012b6dc9bf72cfed651f
react was nil for remote reactions.

Reply:
Oddly, even with this change, whenever I react with my "annoying_dog" emoji, and someone else on an Akkoma instance then reacts on the same reaction, a faux "annoying_dog@akkoma.instance.example" reaction is created (that will also never be able to go away).

Reply:
And if that faux reaction does not exist... It will actually create a fake new remote custom_emoji.

Reply:
Wait wat

Reply:
That's a huge issue I'd love to fix. I definitely don't want Mastodon to create a new remote custom_emoji that doesn't actually exist in the first place...

Reply:
Also, app/lib/activitypub/activity/emoji_react.rb currently also uses @account.domain to search for custom emoji. I believe this is the culprit.

Reply:
Ah!

Reply:
> And if that faux reaction's emoji does not exist... It will actually create a fake new remote custom_emoji.

Just to clarify: I now have a fake "annoying_dog@akkoma.instance.example" custom emoji in the database. I do not yet understand the mechanism behind why this new remote custom emoji is created on the Mastodon side...

Reply:
Honestly that confuses the hell outta me, because an emoji, which doesn't exist, should not the attached to the payload, unless I'm missing or misunderstand something.

Reply:
Oh no... This seems to be a bigger issue with the way custom emojis work.
It is always assumed the domain of a custom emoji is the domain of the account the custom emoji was first seen in...

Reply:
```suggestion
  def set_custom_emoji
    self.custom_emoji = CustomEmoji.find_by(disabled: false, shortcode: name, domain: custom_emoji.domain) if name.present? && custom_emoji.present?
  end
```

I actually don't think it should set a custom emoji when it is nil, as I see it as violation of the spec. Also it leads to unintended behaviour. It doesn't really matter for announcement reactions which this is based on, since they don't handle remote content, but it does matter here.